Incompatible with Pandoc 2.0+

Pandoc introduced a breaking API change for versions 2.0+ in commit [94b3dac](https://github.com/jgm/pandoc/commit/94b3dacb4ea7e5e99ab62286b13877b92f9391b3#diff-2db348243eee8710717037a8cd14c4d3) -- readers now take Text instead of String. As a result, the filter can no longer be installed using Cabal.

The specific build error is:

```
src/Text/Table/Helper.hs:153:47: error:
    • Couldn't match type ‘[Char]’ with ‘Data.Text.Internal.Text’
      Expected type: Data.Text.Internal.Text
        Actual type: String
    • In the second argument of ‘readMarkdown’, namely ‘s’
      In the expression: readMarkdown o s
      In an equation for ‘read’: read = readMarkdown o s
```
It looks like one of the following would need to be updated:

https://github.com/baig/pandoc-csv2table/blob/cbd4ec9e36ad8e697f46f198bf2949b3d336d86c/pandoc-csv2table.cabal#L42

https://github.com/baig/pandoc-csv2table/blob/cbd4ec9e36ad8e697f46f198bf2949b3d336d86c/src/Text/Table/Helper.hs#L149-L153

--- a/src/Text/Table/Helper.hs
+++ b/src/Text/Table/Helper.hs
@@ -56,6 +56,11 @@ import qualified Text.Pandoc.JSON as J
 -- Local imports
 import Text.Table.Definition
 import Text.Table.Builder
+-- imports for compatibility with Pandoc 2.0+
+#if MIN_VERSION_pandoc(2,0,0)
+import Text.Pandoc (runPure)
+import Data.Text (pack)
+#endif
 
 -- Helper functions to manipulate the Pandoc Document and parse the 
 -- Configuration String.
@@ -156,8 +161,19 @@ tableFromCodeBlock as = readMarkdown' def .
                         mkTable (getAtr "caption" as)
                                 (toAlign $ getAtr "aligns" as)
                                 (isHeaderPresent1 $ getAtr "header" as)
+#if MIN_VERSION_pandoc(2,0,0)
+-- Pandoc now operates on Data.Text instead of String. So we pack the string
+-- before calling readMarkdown.
+-- The function signature of readMarkdown has changed. It now returns a PandocMonad
+-- instead of Pandoc. We run readMarkdown in runPure to indicate we are not doing IO
+-- and handle the output in the same manner as in 1.14 migration.
+readMarkdown' :: ReaderOptions -> String -> J.Pandoc
+readMarkdown' o s = case parsed of
+  (Left _) -> J.Pandoc J.nullMeta []
+  (Right p) -> p
+  where parsed = runPure (readMarkdown o $ pack s)
 
-#if MIN_VERSION_pandoc(1,14,0)
+#elif MIN_VERSION_pandoc(1,14,0)
 readMarkdown' :: ReaderOptions -> String -> J.Pandoc
 readMarkdown' o s = case read of
                       (Left _)  -> J.Pandoc J.nullMeta []

--- a/src/Text/Table/Helper.hs
+++ b/src/Text/Table/Helper.hs
@@ -130,6 +130,18 @@ getAtr a ((at,v):_) | a == at = v
 getAtr a (_:xs)               = getAtr a xs
 getAtr a []                   = ""
 
+-- The conversion happens in three stages
+-- 1. Convert the CSV to an internal table representation
+-- 2. Convert internal table representation to markdown
+-- 3. Convert markdown into Pandoc AST.
+--
+-- Converting from internal table representation to Pandoc AST
+-- via markdown decouples internal table representation from
+-- Pandoc AST changes.
+--
+-- TODO: What is the downside of going directly from CSV to Pandoc
+-- table representation?
+
 -- | Make Pandoc Table from Image Inline 
 tableFromImageInline :: [J.Inline] -> CSV -> J.Pandoc
 tableFromImageInline l = addInlineLabel (removeConfigString l) .

--- a/src/Text/Table/Tablify.hs
+++ b/src/Text/Table/Tablify.hs
@@ -45,6 +45,7 @@ import Data.List        (isSuffixOf)
 import Text.Table.Helper
 
 tablifyCsvLinks :: Block -> IO [Block]
+-- variant 1: Referencing CSV file in Image Links
 #if MIN_VERSION_pandoc(1,16,0)
 -- Image Attr [Inline] Target -- ^ Image:  alt text (list of inlines), target
 tablifyCsvLinks (Para [(Image _ l (f, _))]) | "csv" `isSuffixOf` f = do
@@ -59,9 +60,11 @@ tablifyCsvLinks (Para [(Image l (f, _))]) | "csv" `isSuffixOf` f = do
                        toBlocks .
                        tableFromImageInline l $
                        xss
+-- variant 2 and 3: Fenced Code Blocks
 tablifyCsvLinks b@(CodeBlock (_, cs, as) s) | "table" `elem` cs = do
     let file = getAtr "source" as
     case file of
+      -- variant 2: Referencing CSV file in Fenced Code Blocks
       "" -> case s of
               "" -> return [b]
               _  -> case (parseCSV "" s) of
@@ -70,6 +73,7 @@ tablifyCsvLinks b@(CodeBlock (_, cs, as) s) | "table" `elem` cs = do
                                      toBlocks .
                                      tableFromCodeBlock as $
                                      xss
+      -- variant 3:Including CSV content inside Fenced Code Blocks
       _  -> do
               csv <- parseCSVFromFile file
               case csv of
@@ -78,4 +82,5 @@ tablifyCsvLinks b@(CodeBlock (_, cs, as) s) | "table" `elem` cs = do
                                toBlocks .
                                tableFromCodeBlock as $
                                xss
+-- Return input unchanged in case of no match
 tablifyCsvLinks x = return [x]

--- a/stack.yaml
+++ b/stack.yaml
@@ -1,7 +1,7 @@
 # For more information, see: http://docs.haskellstack.org/en/stable/yaml_configuration.html
 
 # Specifies the GHC version and set of packages available (e.g., lts-3.5, nightly-2015-09-21, ghc-7.10.2)
-resolver: lts-8.13
+resolver: lts-12.10
 
 # Local packages, usually specified by relative directory name
 packages:

